Sources/SourceKitLSP/Swift/SyntaxHighlightingToken.swift

```diff
@@ -0,0 +1,413 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SourceKitD
+import LanguageServerProtocol
+import LSPLogging
+
+/// A ranged token in the document used for syntax highlighting.
+public struct SyntaxHighlightingToken: Hashable {
+  public var start: Position
+  public var length: Int
+  public var kind: Kind
+  public var modifiers: Modifiers
+
+  /// The end of a token. Note that this requires the token to be
+  /// on a single line, which is the case for all tokens emitted
+  /// by parseTokens, however.
```

Comment:
VSCode currently doesn't support multi-line tokens (though other LSP clients may, there is a client capability for that).

Removing the single-line assumption sounds good though, perhaps we could just equip `SyntaxHighlightingToken` with a `Range<Position>` field, assuming it might be multi-line by default, and split them into single-line tokens in `SwiftLanguageServer`, before sending them to the client? (I would vote to keep the `splitToSingleLineTokens` method in `SyntaxHighlightingToken` though)

That might actually make it easier to provide multi-line tokens for clients that support them, since we could just query the `clientCapabilities` there.

Reply:
Hm, defining tokens only in terms of `start` and `end: Position` (or a `Range<Position>`) requires document snapshots in quite a few more places, in particular during encoding and decoding, since both SourceKit and LSP talk about lengths.

Not actually sure what the cleanest solution here would be.